Source/RapyutaSimulationPlugins/Private/Tools/SimulationState.cpp

```diff
@@ -21,212 +22,241 @@
-    FServiceCallback GetEntityStateSrvCallback;
-    FServiceCallback SetEntityStateSrvCallback;
-    FServiceCallback AttachSrvCallback;
-    FServiceCallback SpawnEntitySrvCallback;
-    FServiceCallback SpawnEntitiesSrvCallback;
-    FServiceCallback DeleteEntitySrvCallback;
-    GetEntityStateSrvCallback.BindDynamic(this, &ASimulationState::GetEntityStateSrv);
-    SetEntityStateSrvCallback.BindDynamic(this, &ASimulationState::SetEntityStateSrv);
-    AttachSrvCallback.BindDynamic(this, &ASimulationState::AttachSrv);
-    SpawnEntitySrvCallback.BindDynamic(this, &ASimulationState::SpawnEntitySrv);
-    SpawnEntitiesSrvCallback.BindDynamic(this, &ASimulationState::SpawnEntitiesSrv);
-    DeleteEntitySrvCallback.BindDynamic(this, &ASimulationState::DeleteEntitySrv);
-    ROSServiceNode->AddServiceServer(TEXT("GetEntityState"), UROS2GetEntityStateSrv::StaticClass(), GetEntityStateSrvCallback);
-    ROSServiceNode->AddServiceServer(TEXT("SetEntityState"), UROS2SetEntityStateSrv::StaticClass(), SetEntityStateSrvCallback);
-    ROSServiceNode->AddServiceServer(TEXT("Attach"), UROS2AttachSrv::StaticClass(), AttachSrvCallback);
-    ROSServiceNode->AddServiceServer(TEXT("SpawnEntity"), UROS2SpawnEntitySrv::StaticClass(), SpawnEntitySrvCallback);
-    ROSServiceNode->AddServiceServer(TEXT("SpawnEntities"), UROS2SpawnEntitiesSrv::StaticClass(), SpawnEntitiesSrvCallback);
-    ROSServiceNode->AddServiceServer(TEXT("DeleteEntity"), UROS2DeleteEntitySrv::StaticClass(), DeleteEntitySrvCallback);
+    Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+    DOREPLIFETIME(ASimulationState, EntityList);
+    DOREPLIFETIME(ASimulationState, SpawnableEntityInfoList);
+}
 
-    // add all actors
-#if WITH_EDITOR
-    TArray<AActor*> AllActors;
-    UGameplayStatics::GetAllActorsOfClass(GetWorld(), AActor::StaticClass(), AllActors);
-    UE_LOG(LogTemp, Warning, TEXT("Found %d actors in the scene"), AllActors.Num());
+bool ASimulationState::VerifyIsServerCall(const FString& InFunctionName)
+{
+#if RAPYUTA_SIM_DEBUG
+    // Unclear why this check does not work yet?
+    if (IsNetMode(NM_Client))
+    {
+        UE_LOG(LogRapyutaCore, Fatal, TEXT("[%s] should be called by server only"), *InFunctionName);
+        return false;
+    }
+#endif
+    return true;
+}
+
+void ASimulationState::InitEntities()
+{
+#if RAPYUTA_SIM_DEBUG
+    TArray<AActor*> allActors;
+    UGameplayStatics::GetAllActorsOfClass(GetWorld(), AActor::StaticClass(), allActors);
+    UE_LOG(LogTemp, Warning, TEXT("Found %d actors in the scene"), allActors.Num());
 #endif
+    // add all actors
     for (TActorIterator<AActor> It(GetWorld(), AActor::StaticClass()); It; ++It)
     {
         AActor* actor = *It;
         AddEntity(actor);
     }
+
+    // NOTE: [SpawnableEntityInfoList] is a TArray<> thus replicatable, which is not supported for [SpawnableEntities] as a TMap
+    GetWorld()->GetTimerManager().SetTimer(TimerHandle, this, &ASimulationState::GetSpawnableEntityInfoList, 1.0f, true);

```

Comment:
what this timer for?

Reply:
This one is for ensuring that the spawnable entities are replicated over, since spawnable entities doesnt have a trigger that I could borrow to use for emulating replication. The main reason for this and the EntityList 'replication' setups is due to TMaps being unable to be replicated